### DISCUSSION TITLE
PyInstaller-Converted .exe File Not Saving Output Images as Expected

### DISCUSSION THREAD

**Discussion by haqueshenas** [pyinstaller/pyinstaller#8132] [Help]
After converting my .py code to an .exe file, I encountered an issue where running the .exe file does not save the output images in the designated output folders. My code includes three simple and parallel image processing methods, with images being saved successfully by two methods, but encountering an issue with the third method (only in the .exe version).

The following commands are effectively saving the output images:

```output_image_path = os.path.join(Cropped_images, original_filename)```

or this one in another method:

```
f_out = os.path.join(out_path, "C_" + file)
cropped_img.save(f_out)
```

However, the third method fails to save anything (running the .py version does not exhibit this problem):

```
f_out2 = os.path.join(self.Cropped_images, "C_" + file)
cv2.imwrite(f_out2, cv2.cvtColor(cropped_result, cv2.COLOR_RGB2BGR))
```

I would appreciate any guidance or suggestions on resolving this issue.
Thank you.

**Comment by rokm** [comment]
What are the paths in `Cropped_images`, `out_path`, and `self.Cropped_images`? Are they absolute paths or relative paths? If the latter, you are implicitly relying on current working directory, which might not be where you expect.

**Reply by haqueshenas** [reply]
Thank you. So, is it unlikely that the issue is related to the cv2.imwrite function and the way PyInstaller handles it?

In an earlier section of the code, two subfolders, namely 'Segmented images' and 'Cropped images,' were made within the output folder. The location of this output folder can be determined by the user through the GUI (I think they are absolute paths).
```
def settings():
    global v, in_path, out_path
    in_path = directory1Select.folder_path
    out_path = directory2Select.folder_path
```

Within the problematic section, we have:

```
try:
                    self.Segmented_images = os.path.join(out_path, 'Segmented images')
                    self.Cropped_images = os.path.join(out_path, 'Cropped images')

                    os.makedirs(self.Cropped_images)
                    os.makedirs(self.Segmented_images)
                except FileExistsError:
                    pass
```

**Reply by rokm** [reply]
> Thank you. So, is it unlikely that the issue is related to the cv2.imwrite function and the way PyInstaller handles it?

It should not be, and you can always check with a simple example. For example, the following program

```python
import sys
import cv2
img = cv2.imread(sys.argv[1])
cv2.imwrite(sys.argv[2], cv2.cvtColor(img, cv2.COLOR_RGB2BGR))
``` 

seems to work as expected if I freeze it with PyInstaller. Does it for you?

---

Can you add a `print(output_image_path)`, `print(f_out)`, and `print(f_out2)` to your code, so you can see where the images are being saved?

**Reply by rokm** [reply]
Also, does `cv2.imwrite` perhaps throw an exception (invalid path, invalid image data, etc.) when frozen, and you swallow that exception somewhere?

**Reply by haqueshenas** [reply]
Thanks. As a beginner, I used ChatGPT to have a log file, however, only a blank log was saved. Below is the recommendation I got from GPT:

"After freezing the code using PyInstaller, the printed messages won't be visible in the console or terminal because the standalone executable won't have a console attached to it by default.

To log messages to a file instead of printing them, you can use Python's built-in logging module. Here's a simple way to modify your code to log the messages to a file:

Import the logging module at the beginning of your script:

```
import logging
```
Configure the logging settings before your code execution starts. You can do this right after your imports:

```
logging.basicConfig(filename='processing_log.log', level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')
```
Replace print statements with logging statements. For example, in your entw1 method:

Replace:

```
print(f"Saved Segmented Image: {f_out1}")
print(f"Saved Cropped Image: {f_out2}")
```
With:

```
logging.info(f"Saved Segmented Image: {f_out1}")
logging.info(f"Saved Cropped Image: {f_out2}")
```

**Reply by rokm** [reply]
> As a beginner, I used ChatGPT to have a log file, however, only a blank log was saved.

That's between you and ChatGPT to sort out, isn't it? As far as I'm concerned, this is same as complaining that advice you got at the local pub doesn't work for you...

> After freezing the code using PyInstaller, the printed messages won't be visible in the console or terminal because the standalone executable won't have a console attached to it by default.

Are you building with `--windowed`/`--noconsole`/`-W`? If that's the case, yes, you won't get console attached. But you can build without that flag while you debug, and you will get the console. But you should run the executable from command prompt (`cmd`) or Windows Terminal instead of double-clicking of it (otherwise the console will close when the application finishes).

> logging.basicConfig(filename='processing_log.log', ...

FWIW, that's a relative path again, so you will again be making assumptions about current working directory.

**Comment by haqueshenas** [comment]
Ok. I followed your recommendations. However, nothing was printed on the console during or after running the code. Only the two subfolders were made. Besides, I searched the hard drive, but did not find any output image with the similar name I expected.
```
                            f_out2 = os.path.join(self.Cropped_images, "C_" + file)
                            cv2.imwrite(f_out2, cv2.cvtColor(cropped_result, cv2.COLOR_RGB2BGR))

                            # Print statements for the image paths
                            print(f"Saved Cropped Image: {f_out2}")

 ```

**Reply by rokm** [reply]
If you add a `print("hello")` at the very start of the program, is that printed?

**Reply by haqueshenas** [reply]
Yes!
![Hello](https://github.com/pyinstaller/pyinstaller/assets/43292075/6ef8344f-d128-40c5-a236-ff894d732cd6)

**Reply by rokm** [reply]
OK, so the part with the code with other `print()` statements (you are using `print()` instead of logging framework, right?) is probably not being reached at all. Try adding more `print()` statements to figure out where it stops. 

Is this a GUI application? Perhaps the UI callbacks are swallowing exceptions, which would otherwise crash your application...

**Comment by haqueshenas** [comment]
> OK, so the part with the code with other `print()` statements (you are using `print()` instead of logging framework, right?) is probably not being reached at all. Try adding more `print()` statements to figure out where it stops.
> 
> Is this a GUI application? Perhaps the UI callbacks are swallowing exceptions, which would otherwise crash your application...

I removed all "Try-Exceptions" from this part of the code, so the bellow message was appeared on the console: 

"hello
Exception in Tkinter callback
Traceback (most recent call last):
  File "tkinter\__init__.py", line 1921, in __call__
  File "TEasyQuadrat.py", line 1212, in entw1
  File "lazy_loader\__init__.py", line 74, in __getattr__
  File "importlib\__init__.py", line 126, in import_module
  File "<frozen importlib._bootstrap>", line 1050, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1004, in _find_and_load_unlocked
ModuleNotFoundError: No module named 'skimage.segmentation'"

Should I include the "from skimage import segmentation" in my code? Currently, the .py code is run with no error.

**Reply by haqueshenas** [reply]
Thank you so much! The issue has been resolved by adding "from skimage import segmentation."
